src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -341,45 +362,33 @@ protected virtual TypeSymbolWithAnnotations SubstituteType(AbstractTypeMap typeM
         {
             var newCustomModifiers = typeMap.SubstituteCustomModifiers(this.CustomModifiers);
             var newTypeWithModifiers = typeMap.SubstituteType(this.TypeSymbol, withTupleUnification);
-            bool? newIsNullable = (newTypeWithModifiers.IsNullable != null && this.IsNullable != true) ? newTypeWithModifiers.IsNullable : this.IsNullable;
+            bool newIsAnnotated = this.IsAnnotated || newTypeWithModifiers.IsAnnotated;
 
             if (!TypeSymbolEquals(newTypeWithModifiers, TypeCompareKind.CompareNullableModifiersForReferenceTypes) ||
                 !newTypeWithModifiers.CustomModifiers.IsEmpty ||
-                newIsNullable != this.IsNullable ||
+                newIsAnnotated != this.IsAnnotated ||
                 newCustomModifiers != this.CustomModifiers)
             {
-                if (newTypeWithModifiers.TypeSymbol.IsNullableType())
+                var newIsNullableType = newTypeWithModifiers.TypeSymbol.IsNullableType();
+                if (newIsNullableType || !newIsAnnotated)
                 {
-                    Debug.Assert(newIsNullable == true);
                     if (newCustomModifiers.IsEmpty)
                     {
                         return newTypeWithModifiers;
                     }
-
-                    return TypeSymbolWithAnnotations.Create(newTypeWithModifiers.TypeSymbol, newCustomModifiers.Concat(newTypeWithModifiers.CustomModifiers));
+                    newIsAnnotated = newTypeWithModifiers.IsAnnotated;
                 }
-
-                if (newIsNullable == false)
+                else
                 {
-                    Debug.Assert(newTypeWithModifiers.IsNullable != true);
-                    if (newCustomModifiers.IsEmpty)
+                    if (newCustomModifiers.IsEmpty && newTypeWithModifiers.IsAnnotated == newIsAnnotated)
                     {
                         return newTypeWithModifiers;
                     }
-
-                    return TypeSymbolWithAnnotations.Create(newTypeWithModifiers.TypeSymbol, newCustomModifiers.Concat(newTypeWithModifiers.CustomModifiers));
-                }
-
-                Debug.Assert(newIsNullable != false);
-
-                if (newCustomModifiers.IsEmpty && newTypeWithModifiers.IsNullable == newIsNullable)
-                {
-                    return newTypeWithModifiers;
                 }
-
-                return new NonLazyType(newTypeWithModifiers.TypeSymbol, newIsNullable, newCustomModifiers.Concat(newTypeWithModifiers.CustomModifiers));
+                return new NonLazyType(newTypeWithModifiers.TypeSymbol, nonNullTypes: ((NonLazyType)newTypeWithModifiers).NonNullTypes, isAnnotated: newIsAnnotated, newCustomModifiers.Concat(newTypeWithModifiers.CustomModifiers));
```

Comment:
Nit: consider breaking this line.
Also, I found this method hard to follow. Some comments would help.